RecoLocalMuon/GEMSegment/plugins/ME0SegAlgoRU.cc

```diff
@@ -0,0 +1,464 @@
+	return makeFit(proto_segment);
+}
+
+std::unique_ptr<MuonSegFit> ME0SegAlgoRU::makeFit(const HitAndPositionPtrContainer& proto_segment) const {
+  // for ME0 we take the me0rechit from the proto_segment we transform into Tracking Rechits
+  // the local rest frame is the ME0Chamber
+  MuonSegFit::MuonRecHitContainer muonRecHits;
+  for (auto rh=proto_segment.begin();rh<proto_segment.end(); rh++){
+    ME0RecHit *newRH = (*rh)->rh->clone();
+    newRH->setPosition((*rh)->lp);
+    MuonSegFit::MuonRecHitPtr trkRecHit(newRH);
+    muonRecHits.push_back(trkRecHit);
+  }
+  std::unique_ptr<MuonSegFit> currentFit(new MuonSegFit(muonRecHits));
+  currentFit->fit();
+  return currentFit;
+}
+
+void ME0SegAlgoRU::pruneBadHits(const float maxChi2, HitAndPositionPtrContainer& proto_segment,  std::unique_ptr<MuonSegFit>& fit, const unsigned int n_seg_min) const{
+	while(proto_segment.size() > n_seg_min && fit->chi2()/fit->ndof()  > maxChi2){
+		float maxDev = -1;
+		HitAndPositionPtrContainer::iterator worstHit;
+		for(auto it = proto_segment.begin(); it != proto_segment.end(); ++it){
+			const float dev = getHitSegChi2(fit,*(*it)->rh);
+			if(dev < maxDev) continue;
+			maxDev = dev;
+			worstHit = it;
+		}
+		edm::LogVerbatim("ME0SegAlgoRU") << "[ME0SegAlgoRU::pruneBadHits] pruning one hit-> layer: "<< (*worstHit)->layer <<" eta: "<< (*worstHit)->gp.eta() <<" phi: "<<(*worstHit)->gp.phi()<<" old chi2/dof: "<<fit->chi2()/fit->ndof() << std::endl;
+		proto_segment.erase( worstHit);
+		fit = makeFit(proto_segment);
+	}
+}
+
+float ME0SegAlgoRU::getHitSegChi2(const std::unique_ptr<MuonSegFit>& fit, const ME0RecHit& hit) const{
+	const auto lp = hit.localPosition();
+	const auto le = hit.localPositionError();
+	const float du = fit->xdev(lp.x(),lp.z());
+	const float dv = fit->ydev(lp.y(),lp.z());
+
+	ROOT::Math::SMatrix<double,2,2,ROOT::Math::MatRepSym<double,2> > IC;
+	IC(0,0) = le.xx();
+	IC(1,0) = le.xy();
+	IC(1,1) = le.yy();
+
+	// Invert covariance matrix
+	IC.Invert();
+	return du*du*IC(0,0) + 2.*du*dv*IC(0,1) + dv*dv*IC(1,1);
+}
+
+bool ME0SegAlgoRU::hasHitOnLayer(const HitAndPositionPtrContainer& proto_segment, const unsigned int layer) const {
+  for(const auto* h : proto_segment) if(h->layer == layer) return true;
+  return false;
+}
+
+void ME0SegAlgoRU::compareProtoSegment(std::unique_ptr<MuonSegFit>& current_fit, HitAndPositionPtrContainer& current_proto_segment, const HitAndPosition& new_hit) const {
+	const HitAndPosition * old_hit = 0;
+	HitAndPositionPtrContainer new_proto_segment = current_proto_segment;
+
+	HitAndPositionPtrContainer::const_iterator it;
+	for (auto it = new_proto_segment.begin(); it != new_proto_segment.end();) {
+		if ((*it)->layer == new_hit.layer) {
+			old_hit = *it;
+			it = new_proto_segment.erase(it);
+		} else {
+			++it;
+		}
+	}
+	auto new_fit = addHit(new_proto_segment,new_hit);
+
+	//If on the same strip but different BX choose the closest
+	bool useNew = false;
+	if(old_hit->lp == new_hit.lp ){

```

Comment:
in case you haven't noticed this:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17687/18100/llvm-analysis/report-f59f46.html#EndPath